### PR TITLE
rp2040: Added PIO support

### PR DESCRIPTION
**Pull request by yonsch**
Adds support for PIO for the rp2040 soc. PIO programs are not assembled by Zephyr (currently). They should instead be assembled manually and embedded in source code.
The PIO API should preferably be used to implement drivers. For example, to implement UART over the PIO, it should be implemented as a serial driver, selectable by dts. (this is my suggestion at least).
The PIO API from `pico-sdk` can still be used in application code to implement custom applications.

### CODE REVIEW

File: drivers/serial/uart_rpi_pico_pio.c [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	pio_sm_set_enabled(pio, RX_STATE_MACHINE, true);
+}
+
+static int pio_uart_poll_in(const struct device *dev, unsigned char *c)
+{
+	const struct pio_uart_config *config = dev->config;
+	io_rw_8 *uart_rx_fifo_msb;
+	PIO pio = config->pio;
+
+	uart_rx_fifo_msb = (io_rw_8 *)&pio->rxf[RX_STATE_MACHINE] + 3;
```
Comment (lairdjm) [inline drivers/serial/uart_rpi_pico_pio.c:102]: Comment/details on what's special about the magic number 3?

**Review by nordicjm — COMMENTED** [review]
Not tested it but looks good!

**Review by dcpleung — DISMISSED** [review]

File: drivers/serial/uart_rpi_pico_pio.c (hunk 1 above)
Comment (yonsch) [reply drivers/serial/uart_rpi_pico_pio.c:102]: It's a byte pointer to a 32 bit register, so adding 3 gives the MSB of the register, which is denoted in the variable name. But I can document it if it's not clear enough.

**Comment by mbolivar-nordic** [comment]
@petejohanson would you be able to take a look?

File: drivers/serial/uart_rpi_pico_pio.c (hunk 1 above)
Comment (rivimey) [reply drivers/serial/uart_rpi_pico_pio.c:102]: Just browsing, but yes, the 3 needs documenting. I guess there's no need to be worried about big/little endianness because the PIO is built into the soc, but if it were me I would include a comment including "BYTE_ORDER" just to be clear.

**Comment by lairdjm** [comment]
@yonsch please rebase

**Comment by MrGreensWorkshop** [comment]
Hi, I wanted to thank you to  @yonsch for this PR, its gives wide perspective about PIO usage on ZephyrOS.
I think it's better to split PIO support and having uart driver as an example in documents. Because you can write almost anything with PIO, but we can't put everything into Zephyr OS right?

Please give some insights. Thank you.

its better to have PIO initiator as driver. (It's good to have consts like frequency in rp2040.dtsi)
```
static inline void __program_init(PIO pio, uint sm, uint offset, uint in_pin, uint in_pin_count, uint out_pin, uint out_pin_count, float frequency) {
  // 1. Define a config object
  pio_sm_config config = __program_get_default_config(offset);  // 2. Set and initialize the input pins
  sm_config_set_in_pins(&config, in_pin);
  pio_sm_set_consecutive_pindirs(pio, sm, in_pin, in_pin_count, 1);
  pio_gpio_init(pio, in_pin);  // 3. Set and initialize the output pins
  sm_config_set_out_pins(&config, out_pin, out_pin_count);
  pio_sm_set_consecutive_pindirs(pio, sm, out_pin, out_pin_count, 0);  // 4. Set clock divider
  if (frequency < 2000) {
    frequency = 2000;
  }
  float clock_divider = (float) clock_get_hz(clk_sys) / frequency * 1000;
  sm_config_set_clkdiv(&config, clock_divider);  // 5. Configure input shift register
  // args: BOOL right_shift, BOOL auto_push, 1..32 push_threshold
  sm_config_set_in_shift(&config, true, false, 32);  // 6. Configure output shift register
  // args: BOOL right_shift, BOOL auto_push, 1..32 push_threshold
  sm_config_set_out_shift(&config, true, false, 32);  // 7. Join the ISR & OSR
  // PIO_FIFO_JOIN_NONE = 0, PIO_FIFO_JOIN_TX = 1, PIO_FIFO_JOIN_RX = 2
  sm_config_set_fifo_join(&config, PIO_FIFO_JOIN_NONE);  // 8. Apply the configuration
  pio_sm_init(pio, sm, offset, &config);  // 9. Activate the State Machine
  pio_sm_set_enabled(pio, sm, true);
}
```
taken from : https://medium.com/geekculture/raspberry-pico-programming-with-pio-state-machines-e4610e6b0f29

**Comment by yonsch** [comment]
@MrGreensWorkshop Thanks for your feedback. Do I understand correctly that you are proposing a generic PIO driver instead of a per peripheral one? There were a few reasons for implementing it this way:
1. PIO is most useful for implementing uncommon peripherals such as PS/2 or DVI. Therefore, it was assumed that most users would prefer ready to use peripherals, that could be enabled as if they were "regular" peripherals, instead of having a generic PIO driver which would require some extra work to make into a specific peripheral.
2. Implementing PIO peripherals through specific driver APIs allows users to use those peripherals using existing Zephyr API. For example, now you can use the PIO UART as a console, or write a character with `uart_poll_out`, instead of having to use something like `pio_uart_poll_out` or `pio_write`.
3. We preferred not to bother with distributing `pioasm`, since it was only provided as source code as part of `pico-sdk`. With the current implementation, only the driver's author has to bother with installing the assembler.

And finally, the current implementation still gives access to PIO API, so for special cases a user can still write a custom PIO driver, as part of an out of tree Zephyr driver, or simply as part of the application.

**Comment by MrGreensWorkshop** [comment]
@yonsch, Actually I gave my insights from hardware driver point of view. As user point view, probably users prefer to use  ready to use peripherals and I agree with you. Does anyone have any thoughts?

**Comment by rivimey** [comment]
I will need to use the PIO driver for my project, in my case because I want a high resolution PWM output (the code for which is one of the RPi samples.)

I had presumed the Z PIO driver would mostly be about managing access to specific PIO instances (reservation), and enabling loading of pioasm programs. Once run, a PIO unit should probably be represented as a distinct device -- perhaps the model to contemplate is that PIO is a "bus" like PCIe via which devices can be accessed, rather than a device itself?

It would be nice if the Z PIO driver made access to all the RPi PIO sample devices as simple as saying "that one", so you could load it up with an API call, but perhaps that's a later step?

**Comment by MrGreensWorkshop** [comment]
@yonsch Hi, if it's not too much to ask, Could you provide a sample code how to use this driver.
Thank you.

**Comment by yonsch** [comment]
@MrGreensWorkshop Any sample using uart would work, you just need a dts overlay to set up the PIO as a uart. This snippet worked fine when saved as `samples/hello_world/boards/rpi_pico.overlay`:
```
&pinctrl {
	pio_uart0_default: pio_uart0_default {
		group1 {
			pinmux = <PIO0_P3>;
			input-enable;
			bias-pull-up;
		};
		group2 {
			pinmux = <PIO0_P4>;
		};
	};
};
 
&pio0 {
 	status = "okay";
};

&pio_uart0 {
	status = "okay";
	current-speed = <9600>;
	pinctrl-0 = <&pio_uart0_default>;
	pinctrl-names = "default";
};

/ {
	chosen {
		zephyr,console = &pio_uart0;
		zephyr,shell-uart = &pio_uart0;
	};
};

```

**Comment by dcpleung** [comment]
This needs to be rebased.

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 2]
```diff
@@ -155,6 +155,40 @@
...
+			compatible = "raspberrypi,pico-pio";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x50200000 DT_SIZE_K(4)>;
+			clocks = <&system_clk>;
+			resets = <&reset RPI_PICO_RESETS_RESET_PIO0>;
+			label = "PIO_0";
+			status = "disabled";
+
+			pio_uart0: uart {
```
Comment (MrGreensWorkshop) [inline dts/arm/rpi_pico/rp2040.dtsi:169]: I think this definition not specificly needs to be here,

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 3]
```diff
@@ -155,6 +155,40 @@
...
+			compatible = "raspberrypi,pico-pio";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x50300000 DT_SIZE_K(4)>;
+			clocks = <&system_clk>;
+			resets = <&reset RPI_PICO_RESETS_RESET_PIO1>;
+			label = "PIO_1";
+			status = "disabled";
+
+			pio_uart1: uart {
```
Comment (MrGreensWorkshop) [inline dts/arm/rpi_pico/rp2040.dtsi:186]: same as [below](https://github.com/zephyrproject-rtos/zephyr/pull/44316/files#r931179334)

**Review by MrGreensWorkshop — CHANGES_REQUESTED** [review]

File: boards/arm/rpi_pico/doc/index.rst [hunk 4]
```diff
@@ -73,6 +73,9 @@ hardware features:
    * - PWM
      - :kconfig: `CONFIG_PWM`
      - :dtcompatible: `raspberrypi,pico-pwm`
+   * - UART (PIO)
```
Comment (MrGreensWorkshop) [inline boards/arm/rpi_pico/doc/index.rst:76]: I think PIO should be separated from UART. Since PIO is the other peripheral.

File: boards/arm/rpi_pico/doc/index.rst [hunk 5]
```diff
@@ -133,3 +136,14 @@ Inside gdb, run:
...
+Programmable I/O (PIO)
+**********************
+The RP2040 SoC comes with two PIO periherals. These are two simple
+co-processors that are designed for I/O operations. The PIOs run
+a custom instruction set, generated from a custom assembly language.
+PIO programs are assembled using `pioasm`, a tool provided by Raspberry Pi.
+
+Zephyr does not (currently) assemble PIO programs. Rather, they should be
+manually assembled and embedded in source code. An example of how this is done
+can be found at `drivers/serial/uart_rpi_pico_pio.c`.
```
Comment (MrGreensWorkshop) [inline boards/arm/rpi_pico/doc/index.rst:149]: Its better to add a explanation about PIO UART driver. UART could be assigned any pin. (as long as 2 pins in a row)

File: dts/arm/rpi_pico/rp2040.dtsi (hunk 2 above)
Comment (MrGreensWorkshop) [reply dts/arm/rpi_pico/rp2040.dtsi:169]: It's better to have sample code in /samples/boards/

**Comment by dcpleung** [comment]
If this is still relevant, please rebase.

File: boards/arm/rpi_pico/doc/index.rst [hunk 6]
```diff
@@ -112,6 +112,17 @@ Default Zephyr Peripheral Mapping:
 - ADC_CH2 : P28
 - ADC_CH3 : P29
 
+Programmable I/O (PIO)
```
Comment (dcpleung) [inline boards/arm/rpi_pico/doc/index.rst:115]: Please squash this with the previous commit. This would get rid of that merge artifacts there.

**Review by dcpleung — CHANGES_REQUESTED** [review]

**Comment by yonsch** [comment]
@soburi I've integrated your suggestion to this PR, with some changes. Most notably, I implemented the pin naming a bit differently (instead of `gpio-names`, a phandle array of nodelabels), to avoid the double definition of the pin number. Please tell me what you think

**Review by dcpleung — DISMISSED** [review]

File: include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ * @param wrap Wrap source as specified by the PIO program.
+ * @param ... Comma separated list of PIO instructions.
+ */
+#define RPI_PICO_PIO_DECLARE_PROGRAM(name, wrap_target, wrap, ...)	\
+	static const uint32_t name ## _wrap_target = wrap_target;	\
+	static const uint32_t name ## _wrap = wrap;			\
+	static const uint16_t name ## _program_instructions[] = {	\
+		__VA_ARGS__						\
+	};								\
+	static const struct pio_program name ## _program = {		\
```
Comment (fabiobaltieri) [inline include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h:30]: Should this be called just `name`? (as in, drop the `_program` suffix) That way it's a bit easier to understand that this is what it taken by `pio_add_program` as an argument just by reading the code.

File: samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+&pio1 {
+ 	status = "okay";
+
+	pinctrl-0 = <&pio1_uart_default>;
+	pinctrl-names = "default";
+	state-machine-num = <2>;
+
+	pins = <&pio1_uart_rx>, <&pio1_uart_tx>;
```
Comment (soburi) [inline samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay:23]: Need to specify which index of the pinmux array.
I tried it and it seems to work with the attached code.

[0001-add-pinmux-cell.patch](https://github.com/zephyrproject-rtos/zephyr/files/10902389/0001-add-pinmux-cell.patch)

Comment (soburi) [inline samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay:23]: The right hand side value represent pinmux, not pin.
IMO, The property name should change to `pinmuxs`.

**Review by soburi — CHANGES_REQUESTED** [review]

File: samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay (hunk 8 above)
Comment (yonsch) [reply samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay]: done

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 9]
```diff
@@ -193,6 +193,36 @@
...
+		pio0: pio@50200000 {
+			compatible = "raspberrypi,pico-pio";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x50200000 DT_SIZE_K(4)>;
+			clocks = <&system_clk>;
+			resets = <&reset RPI_PICO_RESETS_RESET_PIO0>;
+			status = "disabled";
+
+			pio0_uart: uart {
```
[the remark this answers is not in the document]
Comment (yonsch) [reply dts/arm/rpi_pico/rp2040.dtsi]: agree, done

File: include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h (hunk 7 above)
Comment (yonsch) [reply include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h:30]: I don't know what it will be useful for, we can always add it later

File: samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay (hunk 8 above)
Comment (yonsch) [reply samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay]: Why is the index needed?

File: include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h (hunk 7 above)
Comment (yonsch) [reply include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h:30]: Oh sorry, I though you meant a getter for the instructions themselves. I'll add it

Comment (yonsch) [reply include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h:30]: done

File: samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay (hunk 8 above)
Comment (soburi) [reply samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay]: The `pinmux` property is an array. There is a possibility to use the second or later elements.

e. g. SPI's pinctrl is usually grouped into OUTPUT and INPUT.
If we implement SPI with PIO, we use all of the OUTPUT group's elements.

https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi#L27-L35

File: dts/bindings/misc/raspberrypi,pico-pio.yaml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    description: |
+      A phandle array of pinmuxes used by the higher level driver. The role of
+      each pinmux is determined by `pinmux-names`. For example, if PIO1 is used
+      as a UART peripheral, `pinmuxs` will be set to `<&pio1_uart_rx>, <&pio1_uart_tx>`.
+
+  pinmux-names:
+    type: string-array
+    description: |
+      The names of the pinmuxes as expected by the higher level driver. For example,
+      if the PIO is used as a UART peripheral, `pinmux-names` will be set to `"rx", "tx"`.
```
Comment (gmarull) [inline dts/bindings/misc/raspberrypi,pico-pio.yaml:38]: -1 on custom pinmux schemes

File: dts/bindings/misc/raspberrypi,pico-pio.yaml [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+compatible: "raspberrypi,pico-pio"
+
+include: [base.yaml, pinctrl-device.yaml, reset-device.yaml]
+
+properties:
+  state-machine:
+    type: array
+    description: |
+      Specifies the id of state machines to use.
+      Persistently use the state machine with the specified Id.
```
Comment (gmarull) [inline dts/bindings/misc/raspberrypi,pico-pio.yaml:15]: if this is an array, better use plural name

File: drivers/serial/uart_rpi_pico_pio.c [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+struct pio_uart_config {
+	const struct device *piodev;
+	const uint32_t tx_pin;
+	const uint32_t rx_pin;
+	uint32_t baudrate;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_tx, 0, 3,
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico_pio.c:30]: it looks like this defines, not declares.

File: drivers/serial/uart_rpi_pico_pio.c [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+	sm_config_set_out_pins(&sm_config, pin_tx, 1);
+	sm_config_set_sideset_pins(&sm_config, pin_tx);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, div);
+	sm_config_set_wrap(&sm_config,
+			   offset + RPI_PICO_PIO_GET_WRAP_TARGET(uart_tx),
+			   offset + RPI_PICO_PIO_GET_WRAP(uart_tx));
+
+	pio_sm_set_pins_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
+	pio_sm_set_pindirs_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico_pio.c:72]: if this is the only thing required for pins, why can't pinctrl driver do this if function is PIO0/1?

File: drivers/serial/uart_rpi_pico_pio.c [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+	0x8020, /*  8: push   block                       */
+		/*  .wrap */
+);
+
+static void pio_uart_tx_init(PIO pio, uint32_t sm, uint32_t pin_tx, float div)
+{
+	uint32_t offset;
+	pio_sm_config sm_config;
+
+	offset = pio_add_program(pio, &RPI_PICO_PIO_GET_PROGRAM(uart_tx));
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico_pio.c:58]: suggestion: maybe `GET` macro could actually include `&`

**Review by gmarull — DISMISSED** [review]

File: drivers/serial/uart_rpi_pico_pio.c (hunk 13 above)
Comment (yonsch) [reply drivers/serial/uart_rpi_pico_pio.c]: There are more than just those two functions that use the pin - e.g. `sm_config_set_out_pins`. It is possible to define the configuration of each state machine in dts and have the generic driver apply that configuration. However, I think this would greatly complicate the dts.

File: dts/bindings/misc/raspberrypi,pico-pio.yaml (hunk 10 above)
Comment (yonsch) [reply dts/bindings/misc/raspberrypi,pico-pio.yaml:38]: Would the suggestion by @soburi [here](https://github.com/zephyrproject-rtos/zephyr/pull/44316#discussion_r1127166096) solve this?

Comment (soburi) [reply dts/bindings/misc/raspberrypi,pico-pio.yaml:38]: @gmarull 
Is this pointing out about that the name `"#pinmuxs-cells"` may use commonly, should not use here?
Does the concern solved if the name change to `pio-pinmuxs`?

**Review by dcpleung — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Review by soburi — DISMISSED** [review]

**Comment by yonsch** [comment]
Added DNM until we have a verdict from the architecture WG regarding a common API with #53748

**Review by soburi — COMMENTED** [review]
@iocapa suggest interesting use case in #56678.

In particular, I think it should consider the possibility of creating multiple instances of PIO program
using multiple state machines.

In this case, it makes sense to specify the SM at the child node.

**Comment by yonsch** [comment]
@iocapa Thanks for your effort on #56678 . As this PR is older and has already gone through some reviews, it will be merged first, as the bare minimum required for using PIO. You can then update your PR to add your new features to the existing implementation.
One suggestion that was integrated into this PR was to move the pinctrl to the child nodes rather than the parent PIO device, allowing multiple devices to be defined under one PIO device.
If any of your other changes would require breaking an API later, please say so now, so that I might fix it before this is merged.
@soburi your review would also be appreciated.

File: drivers/misc/pio_rpi_pico/pio_rpi_pico.c [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+int pio_rpi_pico_allocate_sm(const struct device *dev, size_t *sm)
+{
+	const struct pio_rpi_pico_config *config = dev->config;
+	int retval;
+
+	retval = pio_claim_unused_sm(config->pio, false);
+	if (retval < 0) {
+		return -EAGAIN;
```
Comment (soburi) [inline drivers/misc/pio_rpi_pico/pio_rpi_pico.c:25]: It should better to return -EBUSY.

**Review by soburi — CHANGES_REQUESTED** [review]
I added one request.
The other parts look good to me.

And I think we need to clarify and respond to the pointing out by @gmarull.

https://github.com/zephyrproject-rtos/zephyr/pull/44316#discussion_r1127455558

File: drivers/misc/pio_rpi_pico/pio_rpi_pico.c [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+PIO pio_rpi_pico_get_pio(const struct device *dev)
+{
+	const struct pio_rpi_pico_config *config = dev->config;
+
+	return config->pio;
+}
+
+static int pio_rpi_pico_init(const struct device *dev)
+{
+	/* Nothing to do for now */
```
Comment (soburi) [inline drivers/misc/pio_rpi_pico/pio_rpi_pico.c:41]: nitpick:

It seem misleading comment.
It better to remove it.

**Comment by iocapa** [comment]
@yonsch 

> EDIT: I see that #56678 does not need to add `pinmux-cells`. However, in that manner the driver forces the pinctrl to be defined in a certain way. In the example above from the UART driver, it is assumed that the pinctrl is called `pio0_uart0_default`, and that the tx and rx pins are at index 0. For peripherals with more pins (e.g. SPI) that would mean that a specific ordering would be needed in the pinmux array. My point is - IMO both implementations use a custom pinctrl scheme

Yes, I don't see a way not to have a custom pinctrl scheme. 
The initial idea in #56678 was to force the user in defining 
the pins in a certain way since there's a 2nd stage pinmux at 
the PIO level, but still reusing existing infrastructure to 
make it look "similar" to other drivers.

`DT_RPI_PICO_PIO_PIN_BY_NAME` allows getting a `pinctrl-N` 
instance by name, indexing into it, getting a specific group 
by name and indexing into it's pinmux. It was made this way 
to not force anything on the user, other than documenting 
the approach taken. 

In the current scheme (`DT_PIO_PIN_BY_NAME`) you are 
pretty much required to name each and every pin. 
This will get pretty ugly for a high bit count parallel port, 
and you will still be constrained by PIO internals on how 
those pins could be assigned.

For example, looks like the current PR needs to define a 4 bit 
parallel port in the following way:
```
pio_bus_default: pio_bus_default {
		pio_bus: group1 {
			#pinmux-cells = <1>;
			pinmux = <PIO1_P2>, <PIO1_P3>, <PIO1_P4>, <PIO1_P5>;
			input-enable;
			bias-pull-up;
		};
	};
...
pinctrl-0 = <&pio_bus_default>;
pinctrl-names = "default";
pinmuxs = <&pio_bus 0>, <&pio_bus 1>, <&pio_bus 2>, <&pio_bus 3>;
pinmux-names = "d0", "d1", "d2", "d3";
```
However, you can't really change the order, since in the driver you will 
need the base, which will always be the lowest index (0, d0, <&pio_bus 0>), 
and any potential order change needs to be made in 3 different places (pinmux, pinmuxs, pinmux-names).
```
uint8_t base = DT_INST_PIO_PIN_BY_NAME(inst, d0);
```

#56678 does the same thing, but directly at the `pinmux` level.
In this case, a potential change only needs to be made at the `pinmux` level.
```
pio_bus_default: pio_bus_default {
		pio_bus {
			pinmux = <PIO1_P2>, <PIO1_P3>, <PIO1_P4>, <PIO1_P5>;
			input-enable;
			bias-pull-up;
		};
	};
...
pinctrl-0 = <&pio_bus_default>;
pinctrl-names = "pinctrl_name";
```

and getting it by:
```
uint8_t base = DT_INST_RPI_PICO_PIO_PIN_BY_NAME(inst, pinctrl_name, pio_bus, 0);
```

Another thing that is unclear to me is how you would use `pinmuxs` and `pinmux-names` 
to implement more than 1 pinctrl handle. Looks like you need to add it to `pinmux-names` 
but with different names for the same pin.
```
pinctrl-0 = <&pio_bus_default>;
pinctrl-1 = <&pio_bus_extra>;
pinctrl-names = "default", "extra";
pinmuxs = <&pio_bus_default_pins 0>, <&pio_bus_default_pins 1>, <&pio_bus_extra_pins 0>, <&pio_bus_extra_pins 1>;
pinmux-names = "d0_default", "d1_default", "d0_extra", "d1_extra";
```

So IMO, a PIO based driver should document the pin allocation 
scheme requirements as there are some constraints, and the scheme should be decided
by the child driver in a way that makes sense for that specific driver.

**Comment by iocapa** [comment]
@yonsch 

> If any of your other changes would require breaking an API later, please say so now, so that I might fix it before this is merged.

I don't think you can change the current API to be compatible with #56678, as the changes would break the current implementation.

#56678 cannot really coexist with `PICOSDK_USE_PIO` since you would end up with different paths trying to allocate the same resources. Also i see `PICOSDK_USE_CLAIM` creating unnecessary dependencies since any multicore topics should be solved through zephyr specific APIs.

The PIO parent driver in #56678 was designed just as a resource allocator/interrupt router. All child devices need to use the PIO registers directly.
Technically, the PIO part in picosdk is just a wrapper over basic operations and IMO contains a lot of unnecessary overhead for what is does, also some use-cases will result in runtime issues.

For example, when you allocate SMs to different devices you will need to modify some bits in some registers applicable only to those SMs. This requires atomic operations and `picosdk` functions accessing those registers don't support that. In case of UART, if you have the misfortune to end up in a RMW situation on the same register with another device you could end up in an undefined state.

#56678 uses these to solve that:
```
2.1.2. Atomic Register Access
Each peripheral register block is allocated 4kB of address space, with registers accessed using one of 4 methods,
selected by address decode.
• Addr + 0x0000 : normal read write access
• Addr + 0x1000 : atomic XOR on write
• Addr + 0x2000 : atomic bitmask set on write
• Addr + 0x3000 : atomic bitmask clear on write
```

To summarize, my approach would be to break any dependencies on external rpi code (other that the HW structs). This might prove beneficial in the long run as external SW could change and break existing implementations.

**Comment by yonsch** [comment]
@iocapa
> the scheme should be decided by the child driver in a way that makes sense for that specific driver.

Okay, I'm convinced.

> I don't think you can change the current API to be compatible with https://github.com/zephyrproject-rtos/zephyr/pull/56678, as the changes would break the current implementation.

It will break some of the implementation, but the question is if it will break an API, or at least require massive changes. For example, if we merged this PR and then you'll need to move lots of DTS around, that would be counter productive. However if you're just replacing a call to `pio_add_program` with your own implementation, that would be more of an "incremental improvement".
I agree with you that some HAL functionality could be implemented better inside Zephyr, and in the future we might even get rid of the HAL completely (except for hardware definitions as you mentioned).
I would really like to see some of your improvements added, but it is much easier to review in incremental PRs. This PR is pretty old now and has gone through some reviews, so we'd like to merge it. The question is if you see any fundamental problem that will make it too hard for you in the future.

**Review by soburi — APPROVED** [review]

**Comment by iocapa** [comment]
> It will break some of the implementation, but the question is if it will break an API, or at least require massive changes. For example, if we merged this PR and then you'll need to move lots of DTS around, that would be counter productive. However if you're just replacing a call to `pio_add_program` with your own implementation, that would be more of an "incremental improvement".

#56678 breaks pretty much everything on the source code side:
[rpi_pico_pio.h](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio.h)
[rpi_pico_pio_util.h](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio_util.h)
[rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/drivers/misc/rpi_pico_pio/rpi_pico_pio.c)
[uart_rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/drivers/serial/uart_rpi_pico_pio.c)

So, technically, it's a redesign. On a diff, the only similar thing remaining is `DT_RPI_PICO_PIO_PIN_BY_NAME`.
`pio_rpi_pico_allocate_sm` though it looks similar, is not compatible since there's a use-case where you need to allocate more than 1 state machine in a single call because you need explicit ordering for the `INLINE_OUT_EN / OUT_STICKY` feature.

Another issue is that going with incremental merges will not really work since the UART in #56678 needs the PIO in the same PR. So, it could be an incremental review, but needs to be within a single PR.

As I said in the previous comment, #56678 is not compatible with `pico-sdk`. You could technically have an out-of-tree custom driver and still use the SDK (by manually adjusting the allocation scheme), but you cannot really have a in-tree driver using the SDK.
This break was intentional. While the SDK looks ok and there are some examples, they are not really zephyr driver ready and there are some things that the PIO could do which are not really possible with the SDK.
IMO, adding a new driver should follow the same process as other zephyr drivers do: Read the datasheet / Design / Implement.

So, to summarize:
I'm not against merging the current PR.
Just keep in mind that if #56678 should enter (will be rebased as much as possible), in essence it will be just a big diff on the source side (see the files linked above).

**Comment by carlescufi** [comment]
>I'm not against merging the current PR.

Thanks @iocapa. Will merge this one then, and then we can follow-up with improvements.

**Comment by carlescufi** [comment]
Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR

File: boards/arm/rpi_pico/doc/index.rst [hunk 17]
```diff
@@ -109,6 +112,17 @@ Default Zephyr Peripheral Mapping:
...
+**********************
+The RP2040 SoC comes with two PIO periherals. These are two simple
```
Comment (nordicjm) [inline boards/arm/rpi_pico/doc/index.rst:117]:
```suggestion
The RP2040 SoC comes with two PIO peripherals. These are two simple
```

**Review by nordicjm — COMMENTED** [review]
Looks good, tested it and it works, with a general serial throughput test the error rate is about 15-20% but there isn't flow control and this is a simple demo so that's acceptable.

File: drivers/misc/pio_rpi_pico/pio_rpi_pico.c [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+{
+	const struct pio_rpi_pico_config *config = dev->config;
+
+	return config->pio;
+}
+
+static int pio_rpi_pico_init(const struct device *dev)
+{
+	return 0;
+}
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/misc/pio_rpi_pico/pio_rpi_pico.c:42]: +1, can be removed

File: samples/boards/rpi_pico/uart_pio/prj.conf [hunk 19]
```diff
@@ -0,0 +1 @@
+# nothing here
```
Comment (gmarull) [inline samples/boards/rpi_pico/uart_pio/prj.conf:1]: if nothing, just delete the file?

File: samples/boards/rpi_pico/uart_pio/src/main.c [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+ * Copyright (c) 2023 Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+
+void main(void)
```
Comment (gmarull) [inline samples/boards/rpi_pico/uart_pio/src/main.c:11]: `int`

File: samples/boards/rpi_pico/uart_pio/prj.conf (hunk 19 above)
Comment (nordicjm) [reply samples/boards/rpi_pico/uart_pio/prj.conf:1]: prj.conf files are required even if empty

**Comment by gmarull** [comment]
> Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR

A note here: we should not merge something because it comes first, but evaluate both solutions and merge the one that is better.

**Review by MrGreensWorkshop — APPROVED** [review]
LGTM, all of my concerns were fixed. @yonsch Thank you for your time and effort.

There may be a better solution than this, but in my humble opinion, I think we shouldn't be waiting for those to come for months. (fist PR was made on Mar 30, 2022) Let's give people a chance to use it and give some feedback.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
> > Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR
> 
> A note here: we should not merge something because it comes first, but evaluate both solutions and merge the one that is better.

Agreed, this is not the case here. We took this to the architecture WG, and both PRs were scrutinized. No objections were raised to start with this one and then ask @iocapa and @yonsch to bring features from https://github.com/zephyrproject-rtos/zephyr/pull/56678 in progressively.

**Comment by gmarull** [comment]
note: this has been merged with a broken sample https://github.com/zephyrproject-rtos/zephyr/commit/6c93cbf7b496229172fa9508b561417bf91df0ae. Not all comments were addressed, either. @yonsch please submit a follow-up PR or @iocapa feel free to fix in your PR.

**Comment by yonsch** [comment]
Well there were still some unresolved issues (like moving the driver under gio/swio/whatever) but I guess @iocapa can take it from here

**Comment by iocapa** [comment]
Rebased #56678

Please have a look, and let's continue the discussion there.

Ionut
